### ISSUE TITLE
Output object detection per class mAP using faster rcnn model, classwise has been modified to True，but doesn't work.

### ISSUE THREAD

**Issue by LiuChar1esM** [open-mmlab/mmdetection#11535]
I tried to change the \mmdetection\mmdet\evaluation\metrics\coco_metric.py **classwise: bool = False** to  **classwise: bool = True**, but it doesn't work. I have searched related issues but cannot get the expected help.
My system is windows10.Most likely it's because the code is running from source mmdet 3.1.0, which is why it's invalid for me to change classwise in the code. Is there any way to get the config file to load coco_metric.py from code?

my version:
mmdet 3.1.0
torch 1.13.1 + cu116

model:
faster-rcnn_r50_fpn

dataset:
NWPU

config file:
.\mmdetection\configs\faster-rcnn_NWPU2COCO\faster-rcnn_r50_fpn_1x_NWPU.py

```fp16 = dict(loss_scale=512.)
model = dict(
    type='FasterRCNN',
    data_preprocessor=dict(
        # copy from _base_/model/faster-rcnn_r50-fpn
        type='DetDataPreprocessor',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        bgr_to_rgb=True,
        pad_size_divisor=32),
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=True,
        style='pytorch',
        init_cfg=dict(type='Pretrained', checkpoint='torchvision://resnet50')),
    neck=dict(
        type='ATTFPN',
        in_channels=[256, 512, 1024, 2048],  # 输入的各个stage的通道数
        out_channels=256,    # 输出的特征层的通道数
        num_outs=5),    # 输出的特征层的数量
    rpn_head=dict(
        type='RPNHead',
        in_channels=256,
        feat_channels=256,
        anchor_generator=dict(
            type='AnchorGenerator',
            scales=[8],
            ratios=[0.5, 1.0, 2.0],
            strides=[4, 8, 16, 32, 64]),
        bbox_coder=dict(
            type='DeltaXYWHBBoxCoder',
            target_means=[.0, .0, .0, .0],     # 均值
            target_stds=[1.0, 1.0, 1.0, 1.0]),  # 方差
        loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0), # 是否使用sigmoid来进行分类，如果False则使用softmax来分类
        loss_bbox=dict(type='L1Loss', loss_weight=1.0)),
    roi_head=dict(
        type='StandardRoIHead',      # RoIExtractor类型
        bbox_roi_extractor=dict(
            type='SingleRoIExtractor',
            roi_layer=dict(type='RoIAlign', output_size=7, sampling_ratio=0), # ROI具体参数：ROI类型为ROIalign，输出尺寸为7，sample数为2
            out_channels=256,   # 输出通道数
            featmap_strides=[4, 8, 16, 32]),  # 特征图的步长
        bbox_head=dict(
            type='Shared2FCBBoxHead',    # 全连接层类型
            in_channels=256,            # 全连接层数量
            fc_out_channels=1024,      # 输入通道数
            roi_feat_size=7,        # 输出通道数
            # num_classes=80,修改为NWPU的10类
            num_classes=10,       # 分类器的类别数量
            bbox_coder=dict(
                type='DeltaXYWHBBoxCoder',
                target_means=[0., 0., 0., 0.],
                target_stds=[0.1, 0.1, 0.2, 0.2]),
            reg_class_agnostic=False,
            # 是否采用class_agnostic的方式来预测，class_agnostic表示输出bbox时只考虑其是否为前景，
            # 后续分类的时候再根据该bbox在网络中的类别得分来分类，也就是说一个框可以对应多个类别
            loss_cls=dict(
                type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
            loss_bbox=dict(type='L1Loss', loss_weight=1.0))),


    # model training and testing settings
    train_cfg=dict(
        rpn=dict(
            assigner=dict(
                type='MaxIoUAssigner',  # RPN网络的正负样本划分
                pos_iou_thr=0.7,       # 正样本的iou阈值
                neg_iou_thr=0.3,       # 负样本的iou阈值
                min_pos_iou=0.3,       # 正样本的iou最小值。
                # 如果assign给ground truth的anchors中最大的IOU低于0.3，则忽略所有的anchors，否则保留最大IOU的anchor
                match_low_quality=True,
                ignore_iof_thr=-1), # 忽略bbox的阈值，当ground truth中包含需要忽略的bbox时使用，-1表示不忽略
            sampler=dict(
                type='RandomSampler',  # 正负样本提取器类型
                num=256,               # 需提取的正负样本数量
                pos_fraction=0.5,      # 正样本比例
                neg_pos_ub=-1,         # 最大负样本比例，大于该比例的负样本忽略，-1表示不忽略
                add_gt_as_proposals=False),  # 把ground truth加入proposal作为正样本
            allowed_border=-1,      # 不允许在bbox周围外扩一定的像素, 0允许
            pos_weight=-1,          # 正样本权重，-1表示不改变原始的权重
            debug=False),
        rpn_proposal=dict(
            nms_pre=2000,        # 在nms之前保留的的得分最高的proposal数量
            max_per_img=1000,    # 在nms之后保留的的得分最高的proposal数量
            nms=dict(type='soft_nms', iou_threshold=0.7),  # nms阈值
            min_bbox_size=0),    # 最小bbox尺寸
        rcnn=dict(
            assigner=dict(
                type='MaxIoUAssigner',  # RCNN网络正负样本划分
                pos_iou_thr=0.5,   # 正样本的iou阈值
                neg_iou_thr=0.5,   # 负样本的iou阈值
                min_pos_iou=0.5,
                # 正样本的iou最小值。如果assign给ground truth的anchors中最大的IOU低于0.3，则忽略所有的anchors，否则保留最大IOU的anchor
                match_low_quality=False,
                ignore_iof_thr=-1),  # 忽略bbox的阈值，当ground truth中包含需要忽略的bbox时使用，-1表示不忽略
            sampler=dict(
                type='RandomSampler',  # 正负样本提取器类型
                num=512,    # 需提取的正负样本数量
                pos_fraction=0.25,   # 正样本比例
                neg_pos_ub=-1,   # 最大负样本比例，大于该比例的负样本忽略，-1表示不忽略
                add_gt_as_proposals=True),  # 把ground truth加入proposal作为正样本
            pos_weight=-1,   # 正样本权重，-1表示不改变原始的权重
            debug=False)),

    # val_cfg=dict(type='ValLoop'),

    test_cfg=dict(
        rpn=dict(
            nms_pre=2000,
            max_per_img=1000,
            nms=dict(type='soft_nms', iou_threshold=0.7),
            min_bbox_size=0),
        rcnn=dict(
            # score_thr=0.05,
            score_thr=0.001,
            nms=dict(type='soft_nms', iou_threshold=0.5),
            max_per_img=100))
)

# dataset settings
dataset_type = 'NWPU2COCODataset'  # 数据集类型
data_root = 'F:/paper_study/remote_sensing_image_processing/code/mmdetection/data/NWPU2COCO/'
backend_args = None
train_pipeline = [
    dict(type='LoadImageFromFile', backend_args=backend_args),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', scale=(1333, 800), keep_ratio=True),#这个调低点（200,200）
    dict(type='RandomFlip', prob=0.5),
    dict(type='PackDetInputs')
]
test_pipeline = [
    dict(type='LoadImageFromFile', backend_args=backend_args),
    dict(type='Resize', scale=(1333, 800), keep_ratio=True),
    # If you don't have a gt annotation, delete the pipeline
    dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='PackDetInputs',
        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
                   'scale_factor'))
]
# #修改
# metainfo = {
#     'classes': ('airplane', 'ship', 'storage tank', 'baseball diamond', 'tennis court',
#         'basketball court', 'ground track field', 'harbor', 'bridge', 'vehicle')
#     # 'palette': [
#     #     (220, 20, 60),
#     # ]
# }


train_dataloader = dict(
    batch_size=1,
    num_workers=2,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    batch_sampler=dict(type='AspectRatioBatchSampler'),
    dataset=dict(
        type=dataset_type,
        # metainfo=metainfo,
        data_root=data_root,
        ann_file=data_root + 'train.json',
        data_prefix=dict(img=data_root + 'positive_image_set/'),
        # img_prefix=data_root + 'positive_image_set/',
        filter_cfg=dict(filter_empty_gt=True, min_size=32),
        pipeline=train_pipeline,
        backend_args=backend_args
    ))
val_dataloader = dict(
    batch_size=1,
    num_workers=2,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        # metainfo=metainfo,
        data_root=data_root,
        ann_file=data_root+'test.json',
        data_prefix=dict(img=data_root + 'positive_image_set/'),
        # img_prefix=data_root+'positive_image_set/',
        test_mode=True,
        pipeline=test_pipeline,
        backend_args=backend_args
    ))
test_dataloader = val_dataloader

val_evaluator = dict(
    type='CocoMetric',  # 用于评估检测和实例分割的 AR、AP 和 mAP 的 coco 评价指标
    ann_file=data_root + '/test.json',  # 标注文件路径
    metric='bbox',   # 需要计算的评价指标，`bbox` 用于检测，`segm` 用于实例分割
    format_only=False,
    backend_args=backend_args)
test_evaluator = val_evaluator

# training schedule for 1x
train_cfg = dict(type='EpochBasedTrainLoop', max_epochs=24, val_interval=1)
val_cfg = dict(type='ValLoop')
test_cfg = dict(type='TestLoop')

# learning rate
param_scheduler = [
    # dict(
    #     type='LinearLR', start_factor=0.001, by_epoch=False, begin=0, end=500),
    dict(
        type='MultiStepLR',
        begin=0,
        end=24,
        by_epoch=True,
        milestones=[16, 22],
        gamma=0.1)
]

# optimizer
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=dict(type='SGD', lr=0.00125, momentum=0.9, weight_decay=0.0001)) #lr=0.02,8GPU

# Default setting for scaling LR automatically
#   - `enable` means enable scaling LR automatically
#       or not by default.
#   - `base_batch_size` = (8 GPUs) x (2 samples per GPU).
auto_scale_lr = dict(enable=False, base_batch_size=1)

load_from = 'https://download.openmmlab.com/mmdetection/v2.0/faster_rcnn/faster_rcnn_r50_fpn_1x_coco' \
            '/faster_rcnn_r50_fpn_1x_coco_20200130-047c8118.pth '

#default_runtime.py
default_scope = 'mmdet'

default_hooks = dict(
    timer=dict(type='IterTimerHook'),
    logger=dict(type='LoggerHook', interval=50),
    param_scheduler=dict(type='ParamSchedulerHook'),
    checkpoint=dict(type='CheckpointHook', interval=1),
    sampler_seed=dict(type='DistSamplerSeedHook'),
    visualization=dict(type='DetVisualizationHook'))

env_cfg = dict(
    cudnn_benchmark=False,
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    dist_cfg=dict(backend='nccl'),
)

vis_backends = [dict(type='LocalVisBackend')]
visualizer = dict(
    type='DetLocalVisualizer', vis_backends=vis_backends, name='visualizer')
log_processor = dict(type='LogProcessor', window_size=50, by_epoch=True)

log_level = 'INFO'
load_from = None
resume = False
```

My val_evaluator and test_evaluator use same metric: type = CocoMetric.
![image](https://github.com/open-mmlab/mmdetection/assets/44051326/a63ceab6-2298-456b-bb01-f2a545290bc3)

coco_metric.py modified:
![image](https://github.com/open-mmlab/mmdetection/assets/44051326/9edd2d90-3464-48cc-8b82-560fcc33d17a)

I tried to debug as follows：
1. Because of the custom dataset, I  define .\mmdetection\mmdet\datasets\NWPU2COCO.py，and modified classwise = True. It doesn't work, and exegesis this code， gets same unexcepted results.
![image](https://github.com/open-mmlab/mmdetection/assets/44051326/cf9f6f4f-9bed-4136-9395-fd48132723c1)
2. I deleted F:\paper_study\remote_sensing_image_processing\code\mmdetection\mmdet\evaluation\metrics\__init__.py “from .coco_metric import CocoMetric”, run the "python tools/train.py configs/faster-rcnn_NWPU2COCO/faster-rcnn_r50_fpn_1x_NWPU.py",
but it still output results, it's werid!
![image](https://github.com/open-mmlab/mmdetection/assets/44051326/9ab990f8-36ed-4304-b3bc-27108d9b463d)
3. I added some print to get classwise value. But didn't out classwise valuse.
![image](https://github.com/open-mmlab/mmdetection/assets/44051326/24bbb477-7062-4b7a-a002-f62d15e6137a)

**Environment**

sys.platform: win32
Python: 3.8.17 | packaged by conda-forge | (default, Jun 16 2023, 07:01:59) [MSC v.1929 64 bit (AMD64)]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0: NVIDIA GeForce RTX 2070 SUPER
CUDA_HOME: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.6
NVCC: Cuda compilation tools, release 11.6, V11.6.55
MSVC: 用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.38.33133 版
GCC: n/a
PyTorch: 1.13.1+cu116
PyTorch compiling details: PyTorch built with:
  - C++ Version: 199711
  - MSVC 192829337
  - Intel(R) Math Kernel Library Version 2020.0.2 Product Build 20200624 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 2019
  - LAPACK is enabled (usually provided by MKL)
  - CPU capability usage: AVX2
  - CUDA Runtime 11.6
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_7
5,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.3.2  (built against CUDA 11.5)
  - Magma 2.5.4
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.6, CUDNN_VERSION=8.3.2, CXX_COMPILER=C:/actions-runner/_work/pytorch/pytorch/builder/windows/tmp_bin/sccache-cl.exe, CXX_FLAGS=/DWIN32 /D_WINDOWS /GR /EHsc /w /b
igobj -DUSE_PTHREADPOOL -openmp:experimental -IC:/actions-runner/_work/pytorch/pytorch/builder/windows/mkl/include -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DUSE_FBGEMM -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILE
R_USE_KINETO, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NC
CL=OFF, USE_NNPACK=OFF, USE_OPENMP=ON, USE_ROCM=OFF,

TorchVision: 0.14.1+cu116
OpenCV: 4.8.0
MMEngine: 0.8.4
MMDetection: 3.1.0+f78af77

I'm honoured by your advice~！！！！
